---
title: "RA1_without_BCR"
author: "Kristoffer Grimstad"
date: "2022-10-23"
output: html_document
---

# Packages used
```{r}
suppressPackageStartupMessage({
  library(Seurat)
  library(clustree)
  library(DoubletFinder)
  library(scPred)
  library(dplyr)
  library(stringr)
  library(ggplot2)
  library(xlsx)
})
```

# Quality control and filtering
```{r}
## Read in data
set.seed(1001)
ra1.data = Read10X("/RA1_filtered_feature_bc_matrix") # Read in CellRanger output files
ra_no_ig = CreateSeuratObject(counts = ra1.data, project = "RA1", min.cells = 3, min.features = 200) # Create Seurat object

## Visualize QC metrics
ra_no_ig[["percent.mt"]] = PercentageFeatureSet(ra_no_ig, pattern = "^MT-") # Add % mitochondrial genes expressed as metadata
ra_no_ig[["percent.ribo"]] = PercentageFeatureSet(ra_no_ig, pattern = "^RP[SL]") # Add % ribosomal genes as metadata

VlnPlot(ra_no_ig, group.by = "orig.ident", features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), pt.size = 0.1,ncol = 2) + NoLegend()

ra1_no_ig = ra1_no_ig[!grepl("^IG[HKL]V|^IG[KL]J|^IG[KL]C|^IGH[ADEGM]", rownames(ra1_no_ig)),] # Remove Ig genes

ra_no_ig = subset(ra_no_ig, subset = nFeature_RNA > 700 & nFeature_RNA < 3000 & percent.mt < 11 & percent.ribo >5) # Filter for quality

## Normalization
ra_no_ig = NormalizeData(ra_no_ig, normalization.method = "LogNormalize", scale.factor = 10000) # Log-normalize data
ra_no_ig = FindVariableFeatures(ra_no_ig, selection.method = "vst") # Identify highly variable genes

## Scaling
ra_no_ig.all.genes = rownames(ra_no_ig) # All genes in the data
ra_no_ig = ScaleData(ra_no_ig, features = ra_no_ig.all.genes) # Scale

## Principal component analysis
ra_no_ig = RunPCA(object = ra_no_ig, features = VariableFeatures(object = ra_no_ig)) # Perform PCA
ElbowPlot(ra_no_ig, reduction = "pca", ndims = 50) # Plot PCs against standard deviation

## Clustering
ra_no_ig = FindNeighbors(object = ra_no_ig, dims = 1:25) # Construct SNN graph, using 25 PCs

res = seq.int(0.1, 2.0, 0.1)
for (i in res){
  ra_no_ig = FindClusters(object = ra_no_ig, resolution = i)
}
clustree(ra_no_ig@meta.data, prefix = "RNA_snn_res.")
ra_no_ig = FindClusters(object = ra_no_ig, resolution = 0.8) # Identify clusters

## UMAP
ra_no_ig = RunUMAP(object = ra_no_ig, dims = 1:25) # Run UMAP
UMAPPlot(ra_no_ig, label = T) # Plot clusters in UMAP space
```

# Doublet prediction
```{r}
## Define the expected number of doublet cells
annotations = ra_no_ig@meta.data$seurat_clusters # Define clusters
homotypic.prop = modelHomotypic(annotations) # Model homotypic proportions
nExp = round(ncol(ra_no_ig)* 0.031) # Expect 3.1% doublets for 4 000 cells (overexpectation rather than underexpectation)
nExp_adj = round(nExp*(1-homotypic.prop)) # Adjust expectations
ra_no_ig = doubletFinder_v3(ra_no_ig, pN = 0.25, pK = 0.09, nExp = nExp, PCs = 1:25) # Perform doublet prediction
DF.name = colnames(ra_no_ig@meta.data)[grepl("DF.classification", colnames(ra_no_ig@meta.data))]

cowplot::plot_grid( ncol = 2,
                    DimPlot(ra_no_ig, group.by = "orig.ident") + NoAxes(),
                    DimPlot(ra_no_ig, group.by = DF.name) + NoAxes() + ggtitle("Predicted Doublets") + theme(plot.title = element_text(hjust = 0.5))) # Plot predicted doublets

DimPlot(ra_no_ig, group.by = DF.name) + ggtitle("Predicted Doublets") + theme(plot.title = element_text(hjust = 0.5)) # Plot predicted doublets in UMAP space

## Examine if the predicted doublets have more detected genes in general
VlnPlot(ra_no_ig, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1) +
  ggtitle("Predicted Doublets") + theme(plot.title = element_text(hjust = 0.5))

## Remove the predicted doublets from the data
dim(ra_no_ig[,ra_no_ig@meta.data[,DF.name] == "Doublet"])[2] # Number of doublets detected
ra_no_ig = ra_no_ig[,ra_no_ig@meta.data[,DF.name] == "Singlet"] # Only retain predicted singlets for downstream analyses
```

## Cell type prediction
```{r}
reference = scPred::pbmc_1 # Load the reference

## Run all previous analyses on the reference data
reference = reference %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(verbose = F) %>%
  RunUMAP(dims = 1:25)

DimPlot(reference, group.by = "cell_type", label = TRUE, repel = TRUE) + NoAxes() + ggtitle("Reference") + theme(plot.title = element_text(hjust = 0.5)) # Plot predicted cell types in reference in UMAP space
transfer.anchors = FindTransferAnchors(reference = reference, query = ra_no_ig,
                                        dims = 1:25) # Identify transfer anchors between reference and query
predictions = TransferData(anchorset = transfer.anchors, refdata = reference$cell_type, 
                            dims = 1:25) # Transfer predictions using the transfer anchors
ra_no_ig = AddMetaData(object = ra_no_ig, metadata = predictions) # Add predictions to metadata
DimPlot(ra_no_ig, group.by = "predicted.id", label = T, repel = T) + ggtitle("Predicted Cell Type") +  
  theme(plot.title = element_text(hjust = 0.5)) # Plot cell type predictions for query data in UMAP space
```

#  Filtering and reclustering
## Remove irrelevant cells
ra_no_ig = subset(ra_no_ig, predicted.id == "B cell") # Only retain predicted B cells for downstream analysis

## Single-cell transform
ra_no_ig = SCTransform(object = ra_no_ig) # Perform SCT with default settings

## Principal component analysis
ra_no_ig = RunPCA(object = ra_no_ig, features = VariableFeatures(object = ra_no_ig), assay = "SCT") # Perform principal component analysis
ElbowPlot(ra_no_ig, reduction = "pca", ndims = 50) # Plot PCs against standard deviation

## Clustering
ra_no_ig = FindNeighbors(object = ra_no_ig, dims = 1:35) # Construct SNN graph, using 35 PCs
for (i in res){
  ra_no_ig = FindClusters(object = ra_no_ig, resolution = i)
}

clustree(ra_no_ig@meta.data, prefix = "SCT_snn_res.")
ra_no_ig = FindClusters(object = ra_no_ig, resolution = 0.25)

## UMAP
ra_no_ig = RunUMAP(object = ra_no_ig, dims = 1:35) # Run UMAP
UMAPPlot(ra_no_ig, label = T, pt.size = 0.7, label.size = 7) # Plot clusters in UMAP space
```

# Gene expression analyses
```{r}
## Top-10 upregulated DEGs
ra_no_ig_degs = FindAllMarkers(object = ra_no_ig, test.use = "negbinom", return.thresh = 0.05, only.pos = T) # Identify upregulated DEGs
 
ra_no_ig_degs_top10 = ra_no_ig_degs %>% 
  filter(p_val_adj > 0 & p_val_adj < 0.05) %>%
  group_by(cluster) %>%
  slice_min(p_val_adj, n = 10) # Extract top10 significant DEGs

DoHeatmap(object = ra_no_ig, features = ra_no_ig_degs_top10$gene, raster = F) # Plot top10 DEGs

## Add back count data for BCR-genes
a = CreateSeuratObject(counts = ra1.data, project = "RA1", min.cells = 3, min.features = 200)
b = subset(a, cells = WhichCells(object = ra1_no_ig)) # Only include the cells found in the original data

ra1_no_ig = RenameAssays(ra1_no_ig, SCT = 'No_Ig') # Rename assay
ra1_no_ig[["new.counts"]] = CreateAssayObject(counts = GetAssayData(object = b, slot = "count", assay = "RNA"))
ra1_no_ig[["new.counts"]]

ra1_no_ig = SCTransform(ra1_no_ig, assay = "new.counts")

## Plot select genes
FeaturePlot(ra_no_ig, features = c("NFKBIA", "IGHV3-7", "IGKV4-1", "IGLV2-14", "CD69", "COCH", "CRIP2", "TEX9", "ITGB1"), 
            pt.size = 1.5, order = T, cols = c("lightgrey", "red"))

## Plot distribution of IgM from VDJ-seq data
igm = read.xlsx("RA1_IgM_data_VDJ-seq.xlsx", sheetName = "Sheet1")[,1]

DimPlot(ra_no_ig, cells.highlight = igm, pt.size = 1.5, sizes.highlight = 1.5, order = T, cols = c("lightgrey", "red")) + NoLegend()
```

